### ISSUE TITLE
Routes.draw#direct

### ISSUE THREAD

**Issue by dhh** [rails/rails#22512]
It's a common occurrence that the default model-name-to-route direction is too simple to work for applications with nested or programmatic URLs. It's also common that you need to have programmatic routes that delegate to other routes given the state or class of its parameters.

This is a proposal to remedy both situations by allowing the application to direct how instances of certain classes are routed, overwriting the default direction, and to direct programmatic routes according to the passed values.

First, two examples for directing records:

``` ruby
Rails.application.routes.draw do
  # link_to(recording.title, recording) => recordable_path(recording.bucket, recording)
  direct('Recording') { |recording| url_for [ :recordable, recording.bucket, recording ] }

  # link_to(recording.title, recording) => polymorphic_path([ recording.bucket, recording ])
  direct('Recording') { |recording| url_for [ recording.bucket, recording ] }
end
```

The first example refers to a named route called recordable_url and passes two parameters to it (expanded from the single parameter given as the class instance).

The second example simply takes a single instance and turns it into a polymorphic route with an extracted scope parameter.

Now, a few of examples for directing programmatic urls:

``` ruby
Rails.application.routes.draw do
  # bucketable_url(bucket) => project_url(bucket) or circle_url(bucket)
  # bucketable_url(bucket, anchor: 'celebration') => project_url(bucket, anchor: 'celebration')
  direct(:bucketable) { |bucket| url_for [ bucket.bucketable_type.downcase, bucket ] }

  # recordable_path(recording, filter: 'newest') => bucket_message_path(recording.bucket, recording, filter: 'newest')
  direct(:recordable) { |recording| url_for [ "bucket_#{recording.recordable_name}", recording.bucket, recording ] }

  # commentable_url(recording) => polymorphic_url(recording.parent, anchor: recording.anchor_dom_id)
  direct(:commendable) { |recording| url_for [ recording.parent, anchor: recording.anchor_dom_id ] }

  # readable_url(recording) => commentable_url(recording)
  # readable_path(recording) => bucket_chat_around_line_path(recording.bucket, recording.parent, recording)
  direct(:readable) do |recording|
    case recording.recordable_name
    when 'comment', 'client_reply'
      url_for [ :commentable, recording ]
    when /chat_lines/
      if recording.bucket.circle?
        url_for [ :circle_around_line, recording.bucket, recording ]
      else
        url_for [ :bucket_chat_around_line, recording.bucket, recording.parent, recording ]
      end
    when 'question_answer'
      url_for [ :bucket_question_date, recording.bucket, recording.parent, recording.recordable.group_on, anchor: recording.anchor_dom_id ]
    else
      url_for [ :recordable, recording ]
    end
  end
end
```

All these examples are taken from Basecamp 3 where I've currently declared url + path methods in a app/helper that's then included in ApplicationController. This is an incomplete solution in part because these methods are then not available everywhere the router is, like in the ApplicationMailer.

**Comment by pixeltrix** [comment]
I'm generally :+1: but I have a few points:

1. I think conflating mapping models to url helpers and defining custom url helpers that are available everywhere with a single method `direct` will be confusing for new developers. In my opinion it would be better to keep them separate - like `map` and `helper`, e.g:

    ``` ruby
    map("Recording") { |recording| recordable_path(recording.bucket, recording) }
    helper(:bucketable) { |bucketable| [bucket.bucketable_type.downcase, bucket] }
    ```

2. The above example also illustrates my second point - the `follow` method isn't helping in making your examples clearer and it's seems like that it's essentially an alias to `polymorphic_path`. To me it just appears to be noise - IMO it's better to just return the polymorphic arguments we'd pass to `link_to`, etc.

3. We could also provide a non-block syntax for simple cases which provides a easy way to finally address #1769, e.g:

    ``` ruby
    # config/routes.rb
    map "Basket", :basket

    # app/views/baskets/show.html.erb
    <%= form_for @basket do %>
    ```

4. I think it's better to use strings for the model names rather than the class names directly - routes are loaded pretty early and could trigger autoloading problems. It's also the established pattern in Rails with the `:class_name` option for associations in Active Record.

5. The model -> url mapping could also be used to cache polymorphic url generation in the simple case, rather than regenerating the url helper name every time you call `link_to product.title, product`.

> All these examples are taken from Basecamp 3 where I've currently declared url + path methods in a app/helper that's then included in ApplicationController. This is an incomplete solution in part because these methods are then not available everywhere the router is, like in the ApplicationMailer.

One trick I've done in the past for this is to [just use `module_eval` on the named routes module][1]:

```
Rails.application.routes.named_routes.module.module_eval do
  def bucketable_url(bucket, options = {})
    polymorphic_url([bucket.bucketable_type.downcase, bucket, options])
  end
end
```

However that may be broken on master as we've moved the internals around.

[1]: https://github.com/rails/rails/issues/4394#issuecomment-5998326

**Comment by dhh** [comment]
1. I think that's fair debate, though my thinking was that these are basically doing the same thing from different angles: They're working as direction pointers from something that's not a final route to something that is a final route.
2. I use follow to make the case of looking up runtime routes like `"bucket_#{recording.recordable_name}"`. Otherwise you need to use something like public_send which is nasty. Also, I use it because we don't want to define url or path, since these declarations should produce both.
3. 👍
4. 👍
5. 👍

Ah, interesting hack! Hope that I won't need to go there and can just skip straight to this ;).

Is this something you're interested/have time to work on for beta1, or should we canvas for volunteers?

**Comment by matthewd** [comment]
`follow` is there to paper over the difference between `_path` and `_url`, and to pass along any additional parameter hash.

I think the non-block form is a bit too narrow of a use case to warrant its own syntax -- "take any given object, then just route to this static helper" seems quite a rare requirement. And we should be able to make the arrangements inside `resource` itself, rendering the spelling irrelevant, shouldn't we?

**Comment by dhh** [comment]
@matthewd, you mean like:

```ruby
resources :projects do
  resources :posts do
    direct "Post" # => direct("Post") { |post| follow [ [post.project post ] }
  end
end
```

?

**Comment by matthewd** [comment]
@dhh for https://github.com/rails/rails/issues/1769 (singular resources):

```ruby
resource :company
```

could automatically:

```ruby
direct("Company") { follow :company }
```

... so that `form_for(@company)` will correctly submit to `/company` instead of passing the model along to `company_path` (which isn't expecting it).

**Comment by dhh** [comment]
Ah, I see, yes!

**Comment by pixeltrix** [comment]
@dhh I can work on a spike of it this week - there's definitely a few edge cases, for example what happens when the instance is part of a polymorphic array, e.g. `[:admin, @company]` - current behaviour is to generate `admin_companies_path` if `@company.persisted?` is false but `admin_company_path` if true. Would `direct` only apply to the simple case like `form_for @company` ?

**Comment by pixeltrix** [comment]
@matthewd 
> follow is there to paper over the difference between _path and _url, and to pass along any additional parameter hash.

@dhh
> I use follow to make the case of looking up runtime routes like "bucket_#{recording.recordable_name}". Otherwise you need to use something like public_send which is nasty

Both of these concerns can be address by passing the block result to `url_for` - if the block returns a string, it's treated as a fully formed url, a symbol is treated as a url helper name (path or url dependent on context) and an instance or array is passed to `polymorphic_url`. No need to mess with `public_send`.

By using `url_for` on the result you can also return an options hash of `:controller`, `:action`, etc. You can also pass options to `polymorphic_url` by making it the last element of the array.

**Comment by pixeltrix** [comment]
> I think the non-block form is a bit too narrow of a use case to warrant its own syntax -- "take any given object, then just route to this static helper" seems quite a rare requirement.

A non-block form could be optimised - url generation is a hot spot and not having to evaluate a block seems like it could offer a performance boost.

**Comment by dhh** [comment]
@pixeltrix Yeah, my thought was indeed that the direct("Recording") would only be for the `form_for @company` case.

Are you saying that `url_for("bucket_#{bucket.bucketable.name}")`, without a url/path, will do the right thing? Or you're saying we'd change it such that it would? In either case, then that sounds good. Except we also need to remember to pass options, and we don't want to do that manually in the DSL.

**Comment by pixeltrix** [comment]
> Are you saying that url_for("bucket_#{bucket.bucketable.name}"), without a url/path, will do the right thing?

Yes, but it either needs to be a symbol or an array, e.g:

``` ruby
url_for(:"bucket_#{bucket.bucketable.name}")
```

or

``` ruby
url_for(["bucket_#{bucket.bucketable.name}"])
```

A string is treated as url because of `link_to '/foo'`. The benefit of the latter syntax is that you can pass options to the method, e.g.

``` ruby
url_for(["bucket_#{bucket.bucketable.name}", { anchor: dom_id }])
```

By passing the result of the block to `url_for` we can leverage the existing implementation and learning the new feature will be simpler because it's the same as `link_to`, etc.

**Comment by dhh** [comment]
Very nice! Then we don't need the follow part. Sticking with url_for sounds
good to me.

**Comment by dhh** [comment]
(Removing milestone as the PR has it)

**Comment by matthewd** [comment]
Implemented in #23138
